### PR TITLE
AWS: Support shared tag `kubernetes.io/cluster/<clusterid>`

### PR DESCRIPTION
**Pull request by justinsb**
We recognize an additional cluster tag:

kubernetes.io/cluster/<clusterid>

This now allows us to share resources, in particular subnets.

In addition, the value is used to track ownership/lifecycle.  When we
create objects, we record the value as "owned".

We also refactor out tags into its own file & class, as we are touching
most of these functions anyway.

```release-note
AWS: Support shared tag `kubernetes.io/cluster/<clusterid>`
```

### CODE REVIEW

**Comment by justinsb** [comment]
Marking as WIP ...

* [ ] need to make sure we like the tags (and integrate with the feature / documentation work).
* [ ] I think we probably only add the legacy tag if the instance has a legacy tag.  Thus tools can eventually stop adding the legacy tags and they will go away.

We're also laying the groundwork for additional tags on EBS volumes & ELBs here, as this has been frequently requested for billing tagging purposes.

**Comment by justinsb** [comment]
The upgrade is (as I feared) going to be really tough (particularly partial upgrades) so for now we always recognize new & legacy tags (it doesn't cost us much anyway once things have moved to the new tags), and we tag with the new tag always, and the legacy tag only if our instance has a legacy tag.

Thus we will gradually move to only new tags, and the legacy tag matching will become a no-op.

So I think this is ready for review @zmerlynn

**Comment by justinsb** [comment]
@k8s-bot unit test this
@k8s-bot gce etcd3 e2e test this

**Comment by justinsb** [comment]
@k8s-bot node e2e test this

File: pkg/cloudprovider/providers/aws/tags.go [hunk 1]
```diff
@@ -0,0 +1,256 @@
+/*
+Copyright 2014 The Kubernetes Authors.
```
Comment (zmerlynn) [inline pkg/cloudprovider/providers/aws/tags.go]: nit: Copyright 2017

**Review by zmerlynn — APPROVED** [review]
LGTM except a nit. Self-apply /lgtm after that and let's get this tested.

**Comment by justinsb** [comment]
Fixed nit - applying LGTM - thanks!

File: pkg/cloudprovider/providers/aws/tags.go [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+	"github.com/golang/glog"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"strings"
+)
+
+// TagNameKubernetesClusterPrefix is the tag name we use to differentiate multiple
+// logically independent clusters running in the same AZ.
+// The tag key = TagNameKubernetesClusterPrefix + clusterID
+// The tag value is an ownership value
+const TagNameKubernetesClusterPrefix = "kubernetes.io/cluster/"
```
Comment (philips) [inline pkg/cloudprovider/providers/aws/tags.go]: Perhaps consider namespacing this to the cloud provider integration? e.g. `aws.kubernetes.io/cluster-id`?

Comment (justinsb) [inline pkg/cloudprovider/providers/aws/tags.go]: It's an AWS tag

Comment (philips) [inline pkg/cloudprovider/providers/aws/tags.go]: Right, but it is set by the Kubernetes AWS cloud integration; not something core in Kubernetes. I don't have a strong feeling but it feels like if we can be more specific it is better with these unstructured tags.

**Pull request merged**